### PR TITLE
add 404 and 500 handlers

### PR DESCRIPTION
**Pull request by jrperritt**
PR to address [deconst/deconst_docs #35](https://github.com/deconst/deconst-docs/issues/35)

### CODE REVIEW

File: layoutservice/api_at.py [hunk 1]
```diff
@@ -32,18 +36,20 @@ def get(self, presented_url, layout_key):
...
-                    {"Content-Type": "application/json"})
+            # If the domain has a 404 layout, return it. Otherwise, return the
+            # default 404 layout.
+            this_404 = os.path.join(self.layouts_dir, domain, '404.hbs')
+            if os.path.exists(this_404):
+                return send_from_directory(self.layouts_dir,
+                                           domain,
+                                           '404.hbs'), 200
+            return send_from_directory(self.layouts_dir,
+                                       '404.hbs'), 404
```
Comment (smashwilson) [inline layoutservice/api_at.py]: :+1: cool, that does work.

File: layoutservice/api_at.py [hunk 2]
```diff
@@ -32,18 +36,20 @@ def get(self, presented_url, layout_key):
...
-            return (json.dumps({"error": message}),
-                    404,
-                    {"Content-Type": "application/json"})
+            # If the domain has a 404 layout, return it. Otherwise, return the
+            # default 404 layout.
+            this_404 = os.path.join(self.layouts_dir, domain, '404.hbs')
+            if os.path.exists(this_404):
+                return send_from_directory(self.layouts_dir,
+                                           domain,
+                                           '404.hbs'), 200
```
Comment (smashwilson) [inline layoutservice/api_at.py]: Wait, shouldn't this status code be `404` as well?

File: layoutservice/app.py [hunk 3]
```diff
@@ -41,6 +42,31 @@ def get(self):
...
+        layout_filename = code + '.hbs'
+        this_layout = os.path.join(self.layouts_dir, domain, layout_filename)
+        # First look for the layout in the domain's root directory.
+        if os.path.exists(this_layout):
+            return send_from_directory(os.path.join(self.layouts_dir, domain),
+                                       layout_filename), 200
+        # If not found, look for the layout in the 'layouts' directory.
+        elif os.path.exists(os.path.join(self.layouts_dir, layout_filename)):
+            return send_from_directory(self.layouts_dir,
+                                       layout_filename), 200
```
Comment (smashwilson) [inline layoutservice/app.py]: But, confusingly, these should still be 200s, because they're successfully retrieving an error template.

Right - got it.

Comment (jrperritt) [inline layoutservice/app.py]: :+1:

File: layoutservice/app.py [hunk 4]
```diff
@@ -41,6 +42,31 @@ def get(self):
...
+        # Lastly, return the default 500 layout.
+        else:
+            default_500 = os.path.join(self.layouts_dir, '500.hbs')
+            return send_from_directory(default_500)
```
Comment (smashwilson) [inline layoutservice/app.py]: Should we return a 404 here if `default_500` doesn't exist, either?

File: layoutservice/app.py [hunk 5]
```diff
@@ -69,6 +95,14 @@ def get(self):
...
 app.add_url_rule('/status', view_func=StatusHandler.as_view('status'))
 app.add_url_rule('/version', view_func=VersionHandler.as_view('version'))
+app.add_url_rule('/error', view_func=ErrorHandler.as_view('error'))
+
+
+# Default handler for 500 error codes.
+@app.errorhandler(500)
+def server_error(e):
+    return send_from_directory(app.config['settings']['layouts_dir'],
+                               '500.hbs')
```
Comment (smashwilson) [inline layoutservice/app.py]: :+1: nice

File: layoutservice/app.py (hunk 4 above)
Comment (smashwilson) [inline layoutservice/app.py]: Oh - wait. That's what `send_from_directory` will do anyway. Huzzah!

Comment (jrperritt) [inline layoutservice/app.py]: Exactly :)

**Comment by smashwilson** [comment]
Yeah, this looks ready too, modulo that one HTTP status.

**Comment by smashwilson** [comment]
:ship:

**Pull request merged**